File: langs/java-ddd/src/Robot.java
final class Robot {
    Position position;
    Orientation orientation;
    boolean lost = false;

    Robot(Position p, Orientation o) {
        this.position = p;
        this.orientation = o;
    }

    void turnLeft() {
        orientation = orientation.left();
    }

    void turnRight() {
        orientation = orientation.right();
    }

    Position nextForward() {
        return switch (orientation) {
            case N -> new Position(position.x, position.y + 1);
            case E -> new Position(position.x + 1, position.y);
            case S -> new Position(position.x, position.y - 1);
            case W -> new Position(position.x - 1, position.y);
        };
    }
}

File: ocaml-martian-ddd/robot.ml
open Position
open Orientation

type robot = { mutable position: position; mutable orientation: orientation; mutable lost: bool }

let create p o = { position=p; orientation=o; lost=false }
let turn_left r = r.orientation <- left r.orientation
let turn_right r = r.orientation <- right r.orientation
let next_forward r = let dx,dy = step r.orientation in { x = r.position.x + dx; y = r.position.y + dy }

